fraud/bad_encoding.go

```diff
@@ -0,0 +1,219 @@
+package fraud
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// Shares contains all shares from row/col
+	// Shares that did not pass verification in rmst2d will be nil
+	// For non-nil shares MerkleProofs are computed
+	Shares []*Share
+	// Position represents the number of row/col where ErrByzantineRow/ErrByzantineColl occurred
+	Position uint8
+	isRow    bool
+}
+
+func CreateBadEncodingFraudProof(
+	ctx context.Context,
+	height uint64,
+	position uint8,
+	isRow bool,
+	eds *rsmt2d.ExtendedDataSquare,
+	roots [][]byte,
+	errShares [][]byte,
+	dag format.NodeGetter,
+) (Proof, error) {
+	shares := make([]*Share, len(errShares))
+	for index, share := range errShares {
+		if share != nil {
+			data := eds.Row(uint(index))
+			if isRow {
+				data = eds.Col(uint(index))
+			}
+
+			tree, err := buildTreeFromLeaves(ctx, data, roots[index], index, dag)
+			if err != nil {
+				return nil, err
+			}
+			proof, err := tree.Tree().Prove(int(position))
+			if err != nil {
+				return nil, err
+			}
+			/*
+				NamespaceID should be replaced with ParitySharesNamespaceID for
+				all shares except Q0(all erasure coded shares).
+				For EDS 4x4 Q0 is [[0;0],[0;1],[1:0],[1;1]].
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+				x	x	x	x
+			*/
+			namespaceID := share[:consts.NamespaceSize]
```

Comment:
+1. IIRC you're not supposed to manipulate the NID of the parity shares _within the share data_.

(Just as an aside, but making use of the type system instead of remembering that the NID is here vs there would be better.)

Reply:
No, It's not prepended twice at this point. NID is copying to be able to validate the share correctly, as VerifyInclusion is required correct NID.